File: typescript/core/test/compaction.test.ts
import { describe, expect, it } from 'vitest';
import { compact, estimateTokens } from '../src/compaction.js';

type Message = Record<string, unknown>;
const msg = (role: string, content: string): Message => ({ role, content });

describe('compaction', () => {
    it('leaves a conversation under budget unchanged', () => {
        const msgs = [msg('system', 'sys'), msg('user', 'hi'), msg('assistant', 'hello')];
        expect(compact(msgs, 8000)).toEqual(msgs);
    });

    it('is disabled when budget is non-positive', () => {
        const msgs = [msg('user', 'x'.repeat(10_000))];
        expect(compact(msgs, 0)).toEqual(msgs);
    });

    it('drops oldest, keeps system + recent, fits budget', () => {
        const big = 'word '.repeat(200);
        const msgs = [
            msg('system', 'you are helpful'),
            msg('user', `OLDEST ${big}`),
            msg('assistant', `old reply ${big}`),
            msg('user', `MIDDLE ${big}`),
            msg('assistant', `mid reply ${big}`),
            msg('user', 'NEWEST question'),
        ];
        const out = compact(msgs, 400);
        expect(out[0].role).toBe('system');
        const contents = out.map((m) => m.content as string).join(' ');
        expect(contents).toContain('NEWEST question');
        expect(contents).not.toContain('OLDEST');
        expect(out.reduce((s, m) => s + estimateTokens(m), 0)).toBeLessThanOrEqual(400);
    });

    it('never starts the kept window on an orphan tool message', () => {
        const big = 'token '.repeat(300);
        const msgs: Message[] = [
            msg('system', 'sys'),
            msg('user', `q ${big}`),
            { role: 'assistant', content: '', tool_calls: [{ id: 'c1', function: { name: 't', arguments: '{}' } }] },
            { role: 'tool', tool_call_id: 'c1', content: `result ${big}` },
            msg('assistant', 'final answer'),
        ];
        const out = compact(msgs, 200);
        const nonSystem = out.filter((m) => m.role !== 'system');
        expect(nonSystem.length).toBeGreaterThan(0);
        expect(nonSystem[0].role).not.toBe('tool');
    });
});

File: go/core/compaction_test.go
package core

import (
	"strings"
	"testing"
)

func msg(role, content string) ChatMessage {
	return ChatMessage{Role: role, Content: content}
}

func sumTokens(ms []ChatMessage) int {
	total := 0
	for _, m := range ms {
		total += estimateTokens(m)
	}
	return total
}

func TestCompactUnderBudgetUnchanged(t *testing.T) {
	ms := []ChatMessage{msg("system", "sys"), msg("user", "hi"), msg("assistant", "hello")}
	out := compact(ms, 8000)
	if len(out) != len(ms) {
		t.Fatalf("want %d messages, got %d", len(ms), len(out))
	}
}

func TestCompactDisabledWhenNonPositive(t *testing.T) {
	ms := []ChatMessage{msg("user", strings.Repeat("x", 10000))}
	out := compact(ms, 0)
	if len(out) != 1 {
		t.Fatalf("compaction should be disabled at budget 0")
	}
}

func TestCompactDropsOldestKeepsSystemAndRecent(t *testing.T) {
	big := strings.Repeat("word ", 200)
	ms := []ChatMessage{
		msg("system", "you are helpful"),
		msg("user", "OLDEST "+big),
		msg("assistant", "old reply "+big),
		msg("user", "MIDDLE "+big),
		msg("assistant", "mid reply "+big),
		msg("user", "NEWEST question"),
	}
	out := compact(ms, 400)
	if out[0].Role != "system" {
		t.Fatalf("system message must be kept first, got %q", out[0].Role)
	}
	var joined string
	for _, m := range out {
		joined += m.Content + " "
	}
	if !strings.Contains(joined, "NEWEST question") {
		t.Fatalf("newest message should survive; got %q", joined)
	}
	if strings.Contains(joined, "OLDEST") {
		t.Fatalf("oldest message should be dropped; got %q", joined)
	}
	if sumTokens(out) > 400 {
		t.Fatalf("compacted result %d tokens exceeds budget 400", sumTokens(out))
	}
}

func TestCompactNeverStartsOnOrphanTool(t *testing.T) {
	big := strings.Repeat("token ", 300)
	ms := []ChatMessage{
		msg("system", "sys"),
		msg("user", "q "+big),
		{Role: "assistant", ToolCalls: []ToolCall{{ID: "c1", Name: "t", Arguments: "{}"}}},
		{Role: "tool", ToolCallID: "c1", Content: "result " + big},
		msg("assistant", "final answer"),
	}
	out := compact(ms, 200)
	var firstNonSystem *ChatMessage
	for i := range out {
		if out[i].Role != "system" {
			firstNonSystem = &out[i]
			break
		}
	}
	if firstNonSystem == nil {
		t.Fatalf("expected at least one non-system message kept")
	}
	if firstNonSystem.Role == "tool" {
		t.Fatalf("kept window must not start on an orphan tool message")
	}
}
